### ISSUE TITLE
Support for Datadog-style Tags

### ISSUE THREAD

**Issue by 56quarters** [tshlabs/cadence#41]
Per #40, it would be nice to offer support for adding tags to metrics. As part of this, perhaps adding some sort of serialization or formatting layer between `StatsdClient` and the `MetricSink` implementations is in order. When / if this is implemented, let's do it in a way that doesn't exclude us from adding support to the [Metrics 2.0](http://metrics20.org/spec/#wire_format) spec that's floating around.

References:
http://docs.datadoghq.com/guides/dogstatsd/#tags 
http://metrics20.org/spec/#wire_format
https://docs.influxdata.com/influxdb/v1.1/write_protocols/line_protocol_tutorial/

/cc @robinst

**Comment by pjenvey** [comment]
What API do you envision for tags?

Duplicating all methods to add e.g. "incr_with_tags" would be one way. With tags being optional a builder API might be nicer but would require a new backwards incompatible interface.

**Comment by 56quarters** [comment]
I actually hadn't thought about the API too much when I was thinking about implementing it. First I wanted to work on being able to add support for tags without doing any extra heap allocations. Probably overkill in retrospect, it would have been better to start with the API and work from there.

If I (or someone) were to implement this I would want it to look something like this, using the `Counted` trait as an example.

```rust
pub trait MetricBuilder<T> {
    fn with_tag(&mut self, key: &str, val: &str) -> MetricBuilder;

    fn send(&mut self) -> MetricResult<T>;
}

pub trait Counted {
    fn incr(&self, key: &str) -> MetricResult<Counter>;

    fn incr_with_tags(&self, key: &str) -> MetricBuilder<T>;

    // rest of the methods...
}

```

Use would then look like...

```rust
client.incr("some.key").unwrap();

// or

client.incr_with_tags("some.key")
    .with_tag("host", "app01.example.com")
    .with_tag("bucket", "A")
    .send()
    .unwrap();

```

I haven't actually tried to implement this API so I'm not sure how feasible it is.

**Comment by 56quarters** [comment]
Fixed in #61
